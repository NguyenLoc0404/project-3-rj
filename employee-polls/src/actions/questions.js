import { _saveQuestion, _saveQuestionAnswer } from "../utils/_DATA";
import { addUserQuestion, addUserAnswer } from "./users";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ANSWER_QUESTION = "ANSWER_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    qid,
    answer,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authUser } = getState(); 

    return _saveQuestion({
      optionOneText,
      optionTwoText,
      author: authUser,
    }).then((question) => {
      dispatch(addQuestion(question)); 
      dispatch(addUserQuestion(question)); 
    });
  };
}

export function handleAnswerQuestion(qid, answer) {
  return (dispatch, getState) => {
    const { authUser } = getState();

    return _saveQuestionAnswer({
      authedUser: authUser,
      qid,
      answer,
    }).then(() =>
      {
        dispatch(answerQuestion({ authedUser: authUser, qid, answer }));
        dispatch(addUserAnswer({ authedUser: authUser, qid, answer }));
      }
    );
  };
}
