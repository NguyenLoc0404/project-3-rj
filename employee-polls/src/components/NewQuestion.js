import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { handleAddQuestion } from "../actions/questions";
import '../css/new-question.css'; 

const NewQuestion = () => {
  const [optionOneText, setOptionOneText] = useState("");
  const [optionTwoText, setOptionTwoText] = useState("");
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleAddQuestion(optionOneText, optionTwoText));
    navigate("/");
  };

  return (
    <div className="container d-flex justify-content-center align-items-center flex-column">
      <h3 className="mb-4">Would You Rather</h3>
      <p className="mb-4">Create Your Own Poll</p>
      <form onSubmit={handleSubmit} className="d-flex flex-column w-100 align-items-center">
        <div className="mb-3 w-80">
          <label className="form-label">First Option</label>
          <input
            className="form-control"
            type="text"
            value={optionOneText}
            onChange={(e) => setOptionOneText(e.target.value)}
            placeholder="Enter Option One Text Here"
          />
        </div>
        <div className="mb-3 w-80">
          <label className="form-label">Second Option</label>
          <input
            className="form-control"
            type="text"
            value={optionTwoText}
            onChange={(e) => setOptionTwoText(e.target.value)}
            placeholder="Enter Option Two Text Here"
          />
        </div>
        <button type="submit" className="btn btn-primary" disabled={!optionOneText || !optionTwoText}>
          Submit
        </button>
      </form>
    </div>
  );
};

export default NewQuestion;
