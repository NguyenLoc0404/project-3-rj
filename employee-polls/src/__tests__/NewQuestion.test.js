import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import NewQuestion from "../components/NewQuestion";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { MemoryRouter } from "react-router-dom";
import rootReducer from "../reducers";
import { handleAddQuestion } from "../actions/questions";

const store = createStore(rootReducer);
const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

jest.mock("../actions/questions", () => ({
  handleAddQuestion: jest.fn(() => ({
    type: "ADD_QUESTION",
    payload: {
      id: "mock-id",
      optionOneText: "Option One",
      optionTwoText: "Option Two",
      author: "mock-author",
    },
  })),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
})); 

test("NewQuestion component snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter> {/*  Wrap NewQuestion component in MemoryRouter */}
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("fireEvent test for NewQuestion", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );

  const optionOneInput = screen.getByPlaceholderText("Enter Option One Text Here");
  const optionTwoInput = screen.getByPlaceholderText("Enter Option Two Text Here");
  const submitButton = screen.getByText("Submit");

  expect(submitButton).toBeDisabled();

  fireEvent.change(optionOneInput, { target: { value: "Option One" } });
  fireEvent.change(optionTwoInput, { target: { value: "Option Two" } });

  expect(submitButton).not.toBeDisabled();

  fireEvent.click(submitButton);
  expect(mockDispatch).toHaveBeenCalledWith(handleAddQuestion("Option One", "Option Two"));
});


test("initial state of form fields", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );

  const optionOneInput = screen.getByPlaceholderText("Enter Option One Text Here");
  const optionTwoInput = screen.getByPlaceholderText("Enter Option Two Text Here");

  expect(optionOneInput.value).toBe("");
  expect(optionTwoInput.value).toBe("");
});

test("should reset input fields after successful submission", async () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );

  const optionOneInput = screen.getByPlaceholderText("Enter Option One Text Here");
  const optionTwoInput = screen.getByPlaceholderText("Enter Option Two Text Here");
  const submitButton = screen.getByText("Submit");

  fireEvent.change(optionOneInput, { target: { value: "Option One" } });
  fireEvent.change(optionTwoInput, { target: { value: "Option Two" } });

  fireEvent.click(submitButton);

  expect(optionOneInput.value).toBe("Option One");
  expect(optionTwoInput.value).toBe("Option Two");
});


test("renders all elements correctly", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );

  expect(screen.getByPlaceholderText("Enter Option One Text Here")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter Option Two Text Here")).toBeInTheDocument();
  expect(screen.getByText("Submit")).toBeInTheDocument();
});

test("should not dispatch action if input is invalid", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NewQuestion />
      </MemoryRouter>
    </Provider>
  );

  const submitButton = screen.getByText("Submit");

  fireEvent.click(submitButton);
  expect(mockDispatch).not.toHaveBeenCalled();
});