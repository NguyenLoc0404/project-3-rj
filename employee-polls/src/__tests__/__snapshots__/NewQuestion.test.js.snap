// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewQuestion component snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="container d-flex justify-content-center align-items-center flex-column"
  >
    <h3
      class="mb-4"
    >
      Would You Rather
    </h3>
    <p
      class="mb-4"
    >
      Create Your Own Poll
    </p>
    <form
      class="d-flex flex-column w-100 align-items-center"
    >
      <div
        class="mb-3 w-80"
      >
        <label
          class="form-label"
        >
          First Option
        </label>
        <input
          class="form-control"
          placeholder="Enter Option One Text Here"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3 w-80"
      >
        <label
          class="form-label"
        >
          Second Option
        </label>
        <input
          class="form-control"
          placeholder="Enter Option Two Text Here"
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
